index.bs

```diff
@@ -290,3 +290,34 @@ Acknowledgements {#ack}
 
 Cure53's [[DOMPURIFY]] is a clear inspiration for the API this document
 describes, as is Internet Explorer's {{window.toStaticHTML()}}.
+
+
+# Appendix A: Integration with Trusted Types {#trustedtypes}
+
+Issue(WICG/sanitizer-api#20): The best way to integrate Trusted Types is still under discussion.
+
+[Trusted Types](https://w3c.github.io/webappsec-trusted-types/dist/spec/) is
+a proposed feature to allow locking down the DOM to prevent cross-site
+scripting attacks. Trusted Types is complementary to the mechanisms proposed
+here. User Agents that support both, this Sanitizer API and Trusted Types,
+benefit from integration between these features.
+
+To allow easy combination of Sanitizer and Trusted Types, the `Sanitizer`
```

Comment:
When will `sanitizeToTrustedHTML` not be exposed or throw an error? Do we always want to expose `sanitizeToTrustedHTML` because we will only allow developers to create a `Sanitizer` with a safe tags and attributes? If so, I think that's a bad idea to not to allow customizability of `Sanitizer` to contain a unknow attributes in a safe list.

My idea of integration with Trusted Types is:
1. `sanitizeToTrustedHTML` will not be exposed or throw an error when the `SanitizerConfig` doesn't match the default known-safe `SanitizerConfig`.
2.  Developers can still use `sanitize` or `sanitizeToString`  with unsafe config, but they would need to use `TrustedTypePolicy` for that (so the input to `sanitize` needs to be `TrustedHTML` in this case).

Allowing unsafe config in Sanitizer API seems important. Because we want the default `SanitizerConfig` to be always safe, we don't want to allow unknow attributes such as `data-x`. However, Developer would want to allow that depending on the thier usecase, if they know that it's safe. Allowing them to use Sanitizer API with their scenario seems like an important scenario to support, otherwise I imagine we won't get enough adoption of this API.

Reply:
What do you mean by "safe"? Both Sanitizer and Trusted Types do not recognize gadgety payloads, as they are application specific. For example, `class` attribute is a gadgety sink in AngularJS, and yet TT will let `div.class` be set with a string, and Sanitizer will let developers to set it - either by default, or by specifying the `SanitizerConfig`. The baseline is "raw" XSS.

If TT gets mechanics to enforce on gadgety sinks (see https://github.com/w3c/webappsec-trusted-types/issues/288 for example), then this should be aligned (i.e. a lax, gadgety sanitizer config should not be a bypass vector for TT configured for gadgets). But we're not there yet. So far the baseline is XSS, so every `SanitizerConfig` should be OK to produce `TrustedHTML`. Am I missing something?

Reply:
I think we're discussing several things at the same time: What is the default-safe config; to what extend can it be overridden; and how does Trusted Types fit in.

Default-safe config: I think there's consensus that the defaults should be safe, but I think without an explicit defintion of what that entails. My understanding was that these would be anything that directly executes script (e.g. <script> text content, event handlers), or things that load scripts (e.g. <script src=..>). This would not include unknown elements. IOW, the default config permits anything unless it's in the default known-bad list.

Config overrides: The way I've currently spec-ed it, the config cannot override the defaults, ever. I initially thought differently, but thought I had heard consensus on this. This does presume that the defaults will be a fairly minimal known-bad lists, because otherwise we'd be restricting the applicability of the sanitizer.

I'm happy with both of those, but I now wonder whether we really have consensus on them.

TT integration: I think this will look different depending on the previous issues. In terms of mechanics, Making the TT check depending on the config makes me a bit unhappy. I view TT as effectively part of a method's signature, and changing the signature depending on implied object state is difficult to explain (and to implement).

------------ 

Here's a use case, and I wonder what people think of it:

Suppose, an app decides that the class= attribute is part of their threat model (to pick up koto's example; but could also be id=, or whatever). The app would like to use TT + Sanitzer to guard against this.

- I presume we can't block class (id, etc.) as part of the Sanitizer default, since it's commonly useful and not inherently an XSS source.
- The app can easily configure sanitizers to accomodate this. (`blockAttributes = { "class": ["*"] }`, or so.) But it cannot force all sanitizers to do so, since there's no mechanism to limit sanitizer creation.
- If every Sanitizer can create TrustedHTML, then it becomes difficult to ascertain that all TrustedHTML instances will be "safe" according to the app's chosen criteria.

What do? Should there be an on/off switch for sanitizeToTrustedHTML after all? Can we fix this on either TT or Sanitizer side? Or is this not a realistic use case?

Reply:
Trusted Types with their `'script'` sink group (the only one existing for now) only enforce on platform XSS sinks. Additional sink support may be added later, but even with this I don't think it would be feasible to allow for re-configuring the semantics of existing platform sinks (like `className` and `class` attribute). The only feasible way is to introduce new sinks (w3c/webappsec-trusted-types#288), maybe even new types for them (https://gist.github.com/koto/1d044f6029ee337beffb4487b80f8b02). 

If authors in the future want to enforce in their custom sinks (and some of those sinks require `TrustedHTML`) then I think `sanitizeToTrustedHTML` would just unconditionallly throw, and these authors would then have to create TT policies that use their tailored sanitizer policies.

Reply:
I don't think this is about new sinks. (Other than Sanitzer itself arguably being one.) If an app handles untrusted input, surely it'd make sense to use Sanitzer + TT to _also_ ensure that additional, app-specific critera are met, like no class attributes in formatted input? 

(I agree TT couldn't safe authors from themselves if elsewhere they assign untrusted input to class attributes directly.)

Reply:
>What do you mean by "safe"? Both Sanitizer and Trusted Types do not recognize gadgety payloads, as they are application specific.

What I think of safe here is only platform specific sinks. If `class` attribute is a gadgety sink in AngularJS, that doesn't matter here, because if TT is enabled, it will throw an TT violation when the gadgety sink will actually hit the platform specific sinks.

>Default-safe config: I think there's consensus that the defaults should be safe, but I think without an explicit defintion of what that entails. My understanding was that these would be anything that directly executes script (e.g. <script> text content, event handlers), or things that load scripts (e.g. <script src=..>). This would not include unknown elements. IOW, the default config permits anything unless it's in the default known-bad list.

I think a Sanitizer with know-bad list is a bad idea, because I've seen many of them failing in the sanitizer space (e.g. WAFs, XSS Auditors/filters, etc) and generally allow list is usually the safer, logical, and practical option. My opinion for Default-safe configs are:

1. Multiple default configs (with all known-good list),  where the weakest default config would be equivalent of DOMPurify (i.e. allow style tag injection). And the weakest config will also be equivalent to TT enforcement (but we will not allow JavaScript URL to links or frames).
2. People can create their own custom Sanitizer with whatever config they want. However, that won't generate a `TrustedHTML`, and therefore they would need to create a Trusted Type policy if they want the output to be a `TrustedHTML`.


Also, I think it's better to discuss these important things in a meeting like WebAppSec, so that everyone is on the same page. I didn't know about most of "consensus" that are being talked here.

Reply:
I don't think option 2 is on the table. The sanitizer API guarantees that no configuration can become XSSy, so every configuration is safe to generate `TrustedHTML`. How this is actually achieved (e.g. whether the config is based on block- or allowlists) should not be relevant for the TT integration. 

+1 to webappsec, I think we need to discuss the anti-xss approach for the platform anyway, as the sanitizer and CSP only is in my opinion not sufficient.

Reply:
>I don't think option 2 is on the table.

Yeah, if that's not on the table, then that's fine for the TT integration 😊 But I'm certain that people will come to Sanitizer API and beg to allow `javascript:void(0)` and all sort of compatibility problems 😊